File: java/sorting/BitonicSortProduction.java
import java.util.Comparator;
import java.util.List;
import java.util.Objects;

public final class BitonicSortProduction {

    private BitonicSortProduction() {
    }

    public static <T extends Comparable<? super T>> void bitonicSortInPlace(List<T> values) {
        bitonicSortInPlace(values, Comparator.naturalOrder());
    }

    public static <T> void bitonicSortInPlace(List<T> values, Comparator<? super T> comparator) {
        Objects.requireNonNull(values, "values");
        Objects.requireNonNull(comparator, "comparator");

        if (values.size() < 2) {
            return;
        }

        if (!isPowerOfTwo(values.size())) {
            throw new IllegalArgumentException("bitonicSortInPlace requires a power-of-two length");
        }

        bitonicSort(values, 0, values.size(), true, comparator);
    }

    private static boolean isPowerOfTwo(int value) {
        return value > 0 && (value & (value - 1)) == 0;
    }

    private static <T> void bitonicSort(
        List<T> values,
        int start,
        int count,
        boolean ascending,
        Comparator<? super T> comparator
    ) {
        if (count <= 1) {
            return;
        }

        int half = count / 2;
        bitonicSort(values, start, half, true, comparator);
        bitonicSort(values, start + half, half, false, comparator);
        bitonicMerge(values, start, count, ascending, comparator);
    }

    private static <T> void bitonicMerge(
        List<T> values,
        int start,
        int count,
        boolean ascending,
        Comparator<? super T> comparator
    ) {
        if (count <= 1) {
            return;
        }

        int half = count / 2;

        for (int index = start; index < start + half; index++) {
            compareAndSwap(values, index, index + half, ascending, comparator);
        }

        bitonicMerge(values, start, half, ascending, comparator);
        bitonicMerge(values, start + half, half, ascending, comparator);
    }

    private static <T> void compareAndSwap(
        List<T> values,
        int left,
        int right,
        boolean ascending,
        Comparator<? super T> comparator
    ) {
        T leftValue = values.get(left);
        T rightValue = values.get(right);
        boolean shouldSwap = ascending
            ? comparator.compare(rightValue, leftValue) < 0
            : comparator.compare(leftValue, rightValue) < 0;

        if (shouldSwap) {
            values.set(left, rightValue);
            values.set(right, leftValue);
        }
    }
}

File: python/sorting/bitonic_sort_production.py
from collections.abc import Callable, MutableSequence
from typing import TypeVar

T = TypeVar("T")


def _is_power_of_two(value: int) -> bool:
    return value > 0 and value & (value - 1) == 0


def _compare_and_swap(
    values: MutableSequence[T],
    left: int,
    right: int,
    ascending: bool,
    key_fn: Callable[[T], object],
) -> None:
    left_key = key_fn(values[left])
    right_key = key_fn(values[right])
    should_swap = right_key < left_key if ascending else left_key < right_key

    if should_swap:
        values[left], values[right] = values[right], values[left]


def _merge(
    values: MutableSequence[T],
    start: int,
    count: int,
    ascending: bool,
    key_fn: Callable[[T], object],
) -> None:
    if count <= 1:
        return

    half = count // 2

    for index in range(start, start + half):
        _compare_and_swap(values, index, index + half, ascending, key_fn)

    _merge(values, start, half, ascending, key_fn)
    _merge(values, start + half, half, ascending, key_fn)


def _sort(
    values: MutableSequence[T],
    start: int,
    count: int,
    ascending: bool,
    key_fn: Callable[[T], object],
) -> None:
    if count <= 1:
        return

    half = count // 2
    _sort(values, start, half, True, key_fn)
    _sort(values, start + half, half, False, key_fn)
    _merge(values, start, count, ascending, key_fn)


def bitonic_sort_in_place(
    values: MutableSequence[T],
    *,
    key: Callable[[T], object] | None = None,
    reverse: bool = False,
) -> MutableSequence[T]:
    if len(values) < 2:
        return values

    if not _is_power_of_two(len(values)):
        raise ValueError("bitonic_sort_in_place requires a power-of-two length")

    key_fn = key or (lambda item: item)
    _sort(values, 0, len(values), not reverse, key_fn)
    return values
